### ISSUE TITLE
Could not reset offsets to earliest to get the stream from beginning for KTable

### ISSUE THREAD

**Issue by srujanakuntumalla** [spring-cloud/spring-cloud-stream-binder-kafka#376]
Hey,

I am using spring cloud kafka binder to read the data to KStream. While reading the data from one of the topic, i need to read from beginning.

I have tried to set kafka offset reset and start offset properties. But, could not find any references.

Could you please help me providing any sample application.yaml to reset the offset, so that i can consume messages from topic from the beginning

Adding application.yaml that i have used:

```
spring.cloud.stream.bindings.input:
  destination: input-topic1
  consumer:
    useNativeDecoding: true
    headerMode: raw
spring.cloud.stream.bindings.output:
  destination: output-topic
  producer:
    useNativeDecoding: true
    headerMode: raw
spring.cloud.stream.bindings.beginningInput:
  destination: beginning-topic
  consumer:
    useNativeDecoding: true
    headerMode: raw
spring.cloud.stream.kafka.streams.bindings.input:
  consumer:
    keySerde: org.apache.kafka.common.serialization.Serdes$StringSerde
    valueSerde: org.apache.kafka.common.serialization.Serdes$StringSerde
spring.cloud.stream.kafka.streams.bindings.output:
  producer:
    keySerde: org.apache.kafka.common.serialization.Serdes$StringSerde
    valueSerde: org.apache.kafka.common.serialization.Serdes$StringSerde
spring.cloud.stream.kafka.streams.bindings.beginningInput:
  consumer:
    keySerde: org.apache.kafka.common.serialization.Serdes$StringSerde
    valueSerde: org.apache.kafka.common.serialization.Serdes$StringSerde
    resetOffsets: true
    startOffset: earliest
spring.cloud.stream.kafka.streams.binder:
  brokers: 127.0.0.1
  zkNodes: 127.0.0.1
  configuration:
    default.key.serde: org.apache.kafka.common.serialization.Serdes$StringSerde
    default.value.serde: org.apache.kafka.common.serialization.Serdes$StringSerde
    commit.interval.ms: 1000
```

I would like to know if there is any error in the configuration i have used

**Comment by PedroAlvarado** [comment]
Per the code below, it seems that it is not possible to set `resetOffsets` via kafka stream binder consumer properties.

https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/blob/master/spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/properties/KafkaStreamsConsumerProperties.java

**Comment by sobychacko** [comment]
Hi @srujanakuntumalla Currently the kafka streams binder does not expose a way to reset the offset per binding target as the regular MessageChannel based binder does. However, you can do this for the entire application by using this global property: `spring.cloud.stream.kafka.streams.binder.configuration.auto.offset.reset: earliest`. The only problem is that if you have multiple input topics, then this setting is applied to all those topics. I created an issue for introducing this capability per binding: https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/issues/377

**Comment by sobychacko** [comment]
Hi @srujanakuntumalla - After talking through this issue with @garyrussell for a little bit, I am realizing that this issue and potential solutions are much more complex than enabling any offset resetting at the binding level. When we enable offset resetting a the topic/binding level, the first question comes to mind is what will happen to any corresponding intermediate/internal topics and state stores created from the topic. We should ideally reset them as well. Kafka makes available a special tool for these kinds of scenarios (Application reset tool). Here is a good article explaining all the gory details for application resetting in Kafka Streams: https://www.confluent.io/blog/data-reprocessing-with-kafka-streams-resetting-a-streams-application

With that said, can you provide a bit more details about your specific use case? For example, it sounds like you want the input KTable to be read from the beginning each time the app is restarted, but not any other topic/s those are bound as KStream. Is that true? Do you see any side effects for seeking the offset on that topic to the beginning such as any corrupted data downstream in intermediate topics, state stores etc.?

We need to have a clear understanding of these issues before tackling solutions for this in the binder.

**Comment by sobychacko** [comment]
As @garyrussell commented [here](https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/issues/377#issuecomment-388110487), the `auto.offset.reset` property is only for new consumer groups (application.id in the case of kafka streams). As soon as an offset is committed, this property is ignored. In the MessageChannel based Kafka binder, if the application sets `resetOffset` to `earliest` or `latest`, we programmatically seekToBegin or end, thus resetting the offset that was committed before. As I noted in the previous comment above, resetting the offset is not that straight forward in the Kafka Streams binder as there are other moving parts.

**Comment by sobychacko** [comment]
Closing this issue. Please follow the suggestions laid out above if you encounter this use case. Thank you!

**Comment by forgetaboutme** [comment]
@sobychacko not sure what you mean by following the suggestions laid out above. I am using `auto.offset.reset: earliest` within
`spring.cloud.stream.kafka.streams.binder.configuration.auto.offsset.reset` and it is not working.

**Comment by sobychacko** [comment]
@forgetaboutme This issue is very old and closed. Could you please start a new GH issue with the same details above? By "suggestions", I meant the ones from @orchesio on May/2018 above, but happy to look into it further, if you can create a new issue with some context around it.

**Comment by garyrussell** [comment]
@forgetaboutme As stated several times above; that property has no meaning (it is ignored by the kafka client) when there is already a committed offset for the consumer group; it only applies the first time a consumer consumes, or if it never commits an offset, or if the offsets have expired (7 days after the last consumer left the group with Brokers 2.1 and above).

**Comment by forgetaboutme** [comment]
@garyrussell is it possible to reset the offsets?

**Comment by garyrussell** [comment]
Please don't ask the same question in multiple places; it is a waste of our time and yours.

Adding my gitter response here in case people find this thread.

>The `KafkaMessageChannelBinder` has another option `resetOffsets` which applies the `auto.offset.reset` semantics to an existing group by seeking the assigned partitions to the beginning or the end (depending on the `auto.offset.reset` option) when they are first assigned after the binding is started. The Kafka Streams binder does not support that option.

You can use the `kafka-consumer-groups.sh` to reset the offsets (when the real consumer(s) are not connected). `--delete` or `--delete-offsets`.

Or you can add code to create your own consumer to reset (seek) the offsets before starting the streams binding.

**Comment by smeana** [comment]
> Please don't ask the same question in multiple places; it is a waste of our time and yours.
> 
> Adding my gitter response here in case people find this thread.
> 
> > The `KafkaMessageChannelBinder` has another option `resetOffsets` which applies the `auto.offset.reset` semantics to an existing group by seeking the assigned partitions to the beginning or the end (depending on the `auto.offset.reset` option) when they are first assigned after the binding is started. The Kafka Streams binder does not support that option.
> 
> > You can use the `kafka-consumer-groups.sh` to reset the offsets (when the real consumer(s) are not connected). `--delete` or `--delete-offsets`.
> 
> > Or you can add code to create your own consumer to reset (seek) the offsets before starting the streams binding.

So, It is not possible to reset the offset to 0, in a streams application?
The binding consumer properties:
- resetOffsets: true
- startOffset: earliest
- autoCommitOffset: false
seems to do not work for me, been needed to change the application name each time I need to reprocess a topic or delete the consumer.
Is any option to do this just using the .yml properties?
